1.0.3 communication edits

andyjohnson opened the pull request:
Check the individual commits for comments.  Nothing too drastic.  I did change "call" to "request" in some places, which I thought more acceptable.  As with anything in these, please point out if I did something wrong.

garemoko commented on xAPI-Communication.md:
```diff
@@ -180,75 +179,64 @@ The lists above are not intended to be exhaustive. See requirements throughout t
 * The LRS MAY choose to provide both HTTP and HTTPS endpoints to support this use case. 
-* The LRS and the Client SHOULD consider the security risks before making the 
-decision to use this scheme.
+* The LRS and the Learning Record Provider SHOULD consider the security risks before making the decision to use this scheme.
```
I think client was right.

garemoko commented:
Checked the first commit. Left a comment.

garemoko commented:
Just realized that my comment is superceeded by a later commit. 

Looking at that section at the end of 1.3, the paragraph starting "There might be cases where there is a requirement for" has a few problems:

* This scenario and the suggested workarounds are only important if using IE 9 or lower. Its totally possible to use a mix of http and https in other browsers. That whole section needs to be under a heading relating to IE9 and lower, or removed. 
* There's a wierd line break before "Strongly consider"
* The security risk isn't in mixing schemes, it's in using http at all.

garemoko commented on xAPI-Communication.md:
```diff
@@ -224,21 +224,15 @@ __Query string parameters__:
 binary data attachments using this syntax. See [4.1.11. Attachments](#attachments) 
 
-* The LRS MUST support the syntax above.
+* The LRS MUST support the syntax above.  See [Appendix C: Cross Domain Request Example](#Appendix3C) for an example. 
```
The added text should be on a new line otherwise it sounds like the example is just for LRSs.

garemoko replied:
Probably we also need a "requirements" heading

garemoko commented on xAPI-Communication.md:
```diff
@@ -224,21 +224,15 @@ __Query string parameters__:
-
-There might be cases where there is a requirement for the Learning Record Provider to support IE8 and IE9 where the 
-Client code is hosted on a different scheme (HTTP or HTTPS) from the LRS. In these cases, proxy is needed to communicate 
+There might be cases where there is a requirement for the Learning Record Provider to support applications or browsers where the 
```
See comment on the PR

garemoko commented on xAPI-Communication.md:
```diff
@@ -274,14 +268,12 @@ multipart/mixed requests neccesarily do include attachments.
 
-2. The Statement is sent to the receiving system using a Content-Type of
-"multipart/mixed". Any attachments are placed at the end of such transmissions.
+2. The Statement is sent using a Content-Type of"multipart/mixed". Any attachments are placed at the end of such transmissions.
```
missing space

garemoko commented on xAPI-Communication.md:
```diff
@@ -295,12 +287,14 @@ results when the attachments filter is false.
+    * SHOULD include a Content-Type parameter in each part's header. For the first part (containing the Statement) this 
+     MUST be "application/json".
+   	* Where parameters have a corresponding property within the Attachment Object (outlined in the table above), and both 
+   	 the parameter and property are specified for a given Attachment, the value of these parameters and properties MUST match. 
```
we've used "attachment" and "Attachment" interchangeably in this section. I forget which is right.

andyjohnson replied:
I started going through this, but really the whole document needs a style scrub, so I stopped.  The final pass will include synchronizing caps, quotes, use of property, etc.

garemoko commented on xAPI-Communication.md:
```diff
@@ -324,26 +318,25 @@ of Statements which contain only attachment Objects with a populated fileUrl.
 contain attachments.
 
-###### Client Requirements
+###### Learning Record Producer Requirements
```
provider!!!!!

andyjohnson replied:
Sigh, and I was so sure I didn't do it that I didn't do a sweep.  Now I did a sweep.

garemoko commented on xAPI-Communication.md:
```diff
@@ -475,6 +467,9 @@ Stores a single Statement with the given id. POST can also be used to store sing
 ###### LRS Requirements
 
+* While an LRS MUST NOT have duplicate Statement Ids due to the uniqueness requirement, it does not have to perform 
+checks at the time Statements are received.  The following requirements are for LRSs that do such a check:
```
These requirements apply to all LRSs whether they check or not - they MUST not make modifications to their state. 

The option not to check is already covered by 'Whether it responds with `409 Conflict` or `204 No Content`'.

We can remove these two added lines.

andyjohnson replied:
I was trying to add clarity, but I see that I may have opened up a hole in the existing logic.  I removed.

garemoko commented on xAPI-Communication.md:
```diff
@@ -484,15 +479,14 @@ Object.
-* If the LRS receives a Statement with an id it already has a Statement for **in the same batch**, it SHOULD*
-verify the received Statement matches the existing one and SHOULD return `400 Bad Request` if they
-do not match. See [Statement comparision requirements](statement-comparision-requirements).
+* If the LRS receives a Statement with an id it already has a Statement for **in the same batch**, it SHOULD* reject the 
```
How about: "If the LRS receives a batch of Statements containing two or more Statements with the same id..."

andyjohnson replied:
Yes, I left the original text, but that's much better.

garemoko commented on xAPI-Communication.md:
```diff
@@ -467,6 +467,8 @@ Stores a single Statement with the given id. POST can also be used to store sing
 ###### LRS Requirements
 
+* The LRS MAY respond before Statements that have been stored are available for retrieval.
+* 
```
extra star

garemoko commented on xAPI-Communication.md:
```diff
@@ -512,18 +514,22 @@ that provide a lot of data to the LRS.
 have limits. See [Alternate Request Syntax](#alt-request-syntax) for more details.
 * The LRS MUST differentiate a POST to add a Statement or to list Statements based on the 
 parameters passed. See [Alternate Request Syntax](#alt-request-syntax) for more details.
+* While an LRS MUST NOT have duplicate Statement Ids due to the uniqueness requirement, it does not have to perform 
```
See comment on earlier commit

garemoko commented on xAPI-Communication.md:
```diff
@@ -512,18 +514,22 @@ that provide a lot of data to the LRS.
+* If the LRS receives a Statement with an id it already has a Statement for, it SHOULD
+verify the received Statement matches the existing one and SHOULD return `409 Conflict` if they
+do not match. See [Statement comparision requirements](statement-comparision-requirements).
+* If the LRS receives a Statement with an id it already has a Statement for **in the same batch**, it SHOULD* reject the 
```
See comment on earlier commit

garemoko commented on xAPI-Communication.md:
```diff
@@ -726,13 +730,13 @@ are known with reasonable certainty to be available for retrieval. This time SHO
 include attachment raw data and MUST report application/json.
 
-* The LRS SHOULD* include a "Last-Modified" header which matches the Stored timestamp of the Statement. 
+* The LRS SHOULD* include a "Last-Modified" header which matches the "Stored" timestamp of the Statement. 
```
assuming "attachment" is correct, this should also be lowercase

andyjohnson replied:
"stored" is going to wreck our quotes usage, because quotes are necessary in many of the cases to ensure people don't think it is an adjective.  Thanks for spotting this error.

garemoko commented on xAPI-Communication.md:
```diff
@@ -768,13 +770,13 @@ Statements are filtered.
+###### Language Filtering Requirements for Canonical Format Statements
 
-* Activity Objects contain Language Map Objects for name, description and interaction components. 
+* Activity Objects contain Language Map Objects within its "name", "description" and various interaction properties. 
```
"interaction components" are a thing. See part 2.

garemoko replied:
'their' instead of 'its'

andyjohnson replied:
Heh, I wasn't denying they existed, I was trying to be more accurate.  "name" and "description" are properties that directly have these Language Map Objects, they are a bit further down in the interaction components, but I didn't want to say "various properties in interaction components", so I went with just "properties".  I'll revert it and let the readers be smart enough to go look in that section for the usage :)

garemoko commented on xAPI-Communication.md:
```diff
@@ -790,31 +792,29 @@ which language entry to include, rather than to the resource (list of Statements
 Clients can identify the presence and statementId of any voided Statements by the target of the voiding Statement. 
-Aside from debugging tools, many Clients will not want to display voiding statements to their
-users and will not display these as part of activity streams and other reports. 
+Aside from debugging tools, many Clients will not want to display voiding Statements to their users and will not display 
```
Learning Record Consumers instead of Clients.

andyjohnson replied:
We agree the first instance is still "Clients" , right?

garemoko replied:
yep

garemoko commented on xAPI-Communication.md:
```diff
@@ -930,31 +927,29 @@ of "application/json", or if either document cannot be parsed as a JSON Object,
 
-* If an AP needs to delete
-a property, it SHOULD use a PUT request to replace the whole document as described below. 
+* If a Learning Record Producer needs to delete a property, it SHOULD use a PUT request to replace 
```
Provider!

garemoko commented on xAPI-Communication.md:
```diff
@@ -1118,7 +1113,7 @@ about the Agent it received in the request.
 		<td>objectType</td>
 		<td>String</td>
-		<td>"Person"</td>
+		<td>Person</td>
```
We use quotes elsewhere to denote that the value of a property is fixed in the scenario being described. e.g. https://github.com/adlnet/xAPI-Spec/blob/1.0.3-Communication-Edits/xAPI-Data.md#details-2

andyjohnson replied:
Gotcha, thanks - forgot about that.

garemoko commented on xAPI-Communication.md:
```diff
@@ -1466,7 +1464,7 @@ If the header precondition in any of the POST or DELETE request cases above fail
 
 * MUST return HTTP status 409 "Conflict".
-* MUST return a response explaining that the consumer SHOULD
+* MUST return a response explaining that the Client SHOULD
```
If it's a PUT request, this is a Learning Record Provider.

andyjohnson replied:
As you can tell, I'm a bit gunshy on being too specific on some of these.  Thanks for clarifying!

garemoko commented on xAPI-Communication.md:
```diff
@@ -1609,19 +1604,18 @@ This set of credentials SHOULD* be used for conformance testing but MAY be delet
-Systems retrieving Statements might then receive responses that include Statements of different
-versions. The version header allows for these version differences to be handled correctly, and 
-to ascertain that no partial or mixed LRS version implementations exist.
+As a result, Clients might receive responses that include Statements of different versions. The version header allows for these 
```
Learning Record Consumers

garemoko commented on xAPI-Communication.md:
```diff
@@ -1912,17 +1901,15 @@ Participation is highly encouraged.
 ######Conversion of Statements created based on version 0.9
 
-A 1.0.0 system converting a Statement created in 0.9 MUST follow the steps below:
+A 1.0.0 Client converting a Statement created in 0.9 MUST follow the steps below:
```
Mostly it's LRSs that convert statements, but it could be other systems.

andyjohnson replied:
I put in a compromise:  Clients (or other system).

garemoko commented on xAPI-Communication.md:
```diff
@@ -1942,28 +1929,25 @@ A 1.0.0 system converting a Statement created in 0.9 MUST follow the steps below
 ######Conversion of Statements created based on version 0.95
 
-A 1.0.0 system converting a Statement created in 0.95 MUST follow the steps below:
+A 1.0.0 Client converting a Statement created in 0.95 MUST follow the steps below:
```
see above

andyjohnson replied:
Compromised.  Realized parenthesis weren't' necessary.  "Client or other system".

garemoko commented:
@andyjohnson finished review. All my comments are showing as being on an outdated diff and are hidden, but most/all of them are still relevant in the latest version.

andyjohnson commented:
I moved some text in 1.3 and added a couple lines regarding the use of the section.  I'm now claiming that support of IE9 and less is not required, but the requirements in the section are relevant IF they are supported.  Since removing the section is also on the table (I wouldn't mind removing it), we have to be careful that we weren't indicating in previous versions that this syntax MUST be supported.

andyjohnson commented:
Thanks @garemoko for the review.  Well done as always.  Can we get a 3rd pair of eyes on this to get it merged?  Sorry I was slow on this.

rswetnam commented on xAPI-Communication.md:
```diff
@@ -512,18 +514,22 @@ that provide a lot of data to the LRS.
 * The LRS MAY respond before Statements that have been stored are available for retrieval.
 * GET Statements MAY be called using POST and form parameters if necessary as query strings 
 have limits. See [Alternate Request Syntax](#alt-request-syntax) for more details.
 * The LRS MUST differentiate a POST to add a Statement or to list Statements based on the 
```
I find this a bit ambiguous.  I think it is trying to say The LRS MUST differentiate POSTs to add Statements from POSTs to list Statements...  ?

garemoko replied:
@rswetnam  How about "The LRS MUST differentiate between a POST to add a Statement or a POST to list Statements based on the "

garemoko commented on xAPI-Communication.md:
```diff
@@ -177,6 +177,8 @@ The lists above are not intended to be exhaustive. See requirements throughout t
 
 ### 1.3 Alternate Request Syntax
 
+__Note__: Unless supporting IE9 or lower, this section is not necessary for implementations of xAPI.
```
This is not true. LRSs *MUST* support this and clients *MAY* use this even when not using IE9 or lower.

garemoko commented on xAPI-Communication.md:
```diff
@@ -189,6 +191,12 @@ the xAPI as described above due to only supporting "GET" and "POST", and not all
 The following describes alternate syntax to use only when unable to use the usual syntax for specific calls due to the 
 restrictions mentioned above. This alternate syntax can also be used to GET Statements due to limits on query string length.  
 
+All requirements for this environment are necessary only if supporting versions of browsers requiring this type of support.
```
Repeated from above: This is not true. LRSs *MUST* support this and clients *MAY* use this even when not using IE9 or lower.

garemoko commented on xAPI-Communication.md:
```diff
@@ -1464,7 +1465,7 @@ If the header precondition in any of the POST or DELETE request cases above fail
 
 * MUST return HTTP status 409 "Conflict".
-* MUST return a response explaining that the Client SHOULD
+* MUST return a response explaining that the Learning Record Producer SHOULD
```
Provider!!!

garemoko commented:
Added some more comments; still showing as on an outdated diff for some reason, but definitely relevant. 

I haven't reviewed to check that all my earlier comments had been acted on.

andyjohnson commented:
So to be clear (probably need a final discussion on the issue), the Section 1.3 on Cross Domain Scripting does indeed need to be supported by LRSs and we cannot remove the section or any of the requirements in version 1.0.X. All comments should be addressed.  Was a bit rushed before to get the last round in.  Thanks for the reviews @garemoko !

oconnetf commented on xAPI-Communication.md:
```diff
@@ -261,70 +267,70 @@ of Statements which contain only attachment Objects with a populated fileUrl.
+* A Learning Record Provider MAY send batches of type "application/json" where every attachment
 Object has a fileUrl, ignoring all requirements based on the "multipart/mixed" format.
-* A Learning Record Producer SHOULD use SHA-256, SHA-384, or SHA-512  to populate the "sha2" property.
+* A Learning Record Provider SHOULD use SHA-256, SHA-384, or SHA-512  to populate the "sha2" property.
```
Extra space after "SHA-512".

oconnetf commented on xAPI-Communication.md:
```diff
@@ -103,17 +103,16 @@
 ## 1.0 Requests
 
-xAPI tracking is done via HTTP Requests from the Activity Provider (client) to the LRS (server).  This 
+xAPI tracking is done via HTTP Requests from the Learning Record Provider (client) to the LRS (server).  This 
```
Two spaces after period.

oconnetf commented:
I tossed in a couple of very minor whitespace issues I noticed.

As long as @garemoko accepts your perspective on 1.3, I'm 👍.

garemoko commented:
![image](https://cloud.githubusercontent.com/assets/1363163/15041627/53bcfe64-12b5-11e6-9c67-99cb595f489b.png)

Pink arrow: A proxy is only needed IF you want to support HTTP to HTTPS requests **from IE 9 or lower**. You can do HTTP to HTTPS (or HTTPS to HTTP!) without a proxy if you use a modern browser. 

Green arrow: paragraph spacing

garemoko commented:
Also, RE the last sentence, the security risk isn't in mixing schemes, it's in using http at all. I think you missed this comment: https://github.com/adlnet/xAPI-Spec/pull/918#issuecomment-214516051

garemoko commented:
Space added in response to https://github.com/adlnet/xAPI-Spec/pull/918#discussion_r60986479 but it's on the wrong side of the quote.

garemoko commented:
Not sure if this is in scope of this PR or not but...
![image](https://cloud.githubusercontent.com/assets/1363163/15041763/5c3769b6-12b6-11e6-9f6c-a9e0de153873.png)

garemoko commented:
Might also be out of scope of this PR, but I'm not sure what the following paragraph adds or even what point it's trying to make:

`As a result, Learning Record Consumers might receive responses that include Statements of different versions. The version header allows for these version differences to be handled correctly, and to ascertain that no partial or mixed LRS version implementations exist.`

I propose we delete it and merge the remaining two sentences in 3.3 Versioning into a single paragraph. 

Also, in that next sentence, can we change "reliably know compatibility" to "remain interoperable"?

garemoko commented:
Aside from the comments just added, looks like all other comments have been actioned (or at least responded to in the case of capitalization). Thanks!

andyjohnson commented:
Agreed on the paragraph getting deleted, made those changes.  In regards to the bullet spacing, it is right down the middle in terms of usage/non-usage.  It is something to look for in the final pass as it is in all documents.  I'm learning towards using spaces as some of the bullet requirements can be entire paragraphs.  I think we'll need a final quick look at the Cross-Domain/HTTP Security section as far as content, but I recommend merging this PR (I went with all of your suggestions) and tackling it separately.  Thanks to you both, @garemoko and @oconnetf for reviewing this PR multiple times.

garemoko commented:
Yeh, agree that the cross scheme paragraph needs more attention; (I didn't expect you to copy my words exactly and some of it is a little rough)

Honestly I'd prefer to tackle that in this PR but I don't mind handling it separately if you prefer so +1

andyjohnson commented:
Thanks @garemoko - it frees me up to do other PRs and have peace of mind, so I'd rather merge it.  I'm travelling now but should be back on this tomorrow.
